File: tools/celery_zstd_experiment/churn.py
#!/usr/bin/env python3
"""
  * producer (async task): calls the real `process_job.delay()` continuously
  * consumer (async task): drives a `celery worker` subprocess. The worker
    is paused with SIGSTOP so the queue builds, then every `--drain-every`
    seconds it is resumed (SIGCONT) for a `--drain-window` burst during which it
    actually BRPOPs + executes tasks (decompressing nodes on read), then paused
    again

Usage:
    python churn.py --produce-rate 1500 --drain-every 20 --drain-window 8
"""

import argparse
import asyncio
import os
import signal
import sys
import uuid
from concurrent.futures import ThreadPoolExecutor

import redis.asyncio as aioredis

from celery_app import process_job

HERE = os.path.dirname(os.path.abspath(__file__))
WORKER_LOG = "/tmp/celery_zstd_worker.log"


_WAREHOUSES = ["us-east-1", "us-west-2", "eu-central-1", "ap-south-1"]
_STATUSES = ["pending", "processing", "shipped", "backordered"]


def make_payload(i, n_items):
    """Structured task payload. With n_items>0 it carries a list of records with a
    fixed schema but realistically varying values (incl. a random per-record id),
    like a real batch task - compressible via the shared structure, not
    artificially so."""
    p = {"user_id": i % 10000, "action": "process", "retries": 0, "queue": "my_queue", "attempt": 1}
    if n_items:
        p["items"] = [
            {
                "sku": f"SKU-{(i * 7 + k * 13) % 1000000:07d}",
                "qty": (k % 9) + 1,
                "price": f"{((i + k) % 5000) / 100:.2f}",
                "warehouse": _WAREHOUSES[(i + k) % 4],
                "status": _STATUSES[(i + k) % 4],
                "line_id": str(uuid.uuid4()),
            }
            for k in range(n_items)
        ]
    return p


def publish_batch(start_i, n, n_items):
    """Synchronously publish n real Celery tasks (runs in a worker thread)."""
    for j in range(n):
        i = start_i + j
        process_job.delay(
            job_id=f"b3e4b923-8a77-4053-aff0-{100000 + i}",
            payload=make_payload(i, n_items),
        )
    return start_i + n


async def _sleep_or_stop(stop, secs):
    try:
        await asyncio.wait_for(stop.wait(), timeout=secs)
    except asyncio.TimeoutError:
        pass
    return stop.is_set()


def _signal_worker(worker, sig):
    """Send a signal to the worker, tolerating an already-dead process."""
    if worker.returncode is None:
        try:
            worker.send_signal(sig)
        except ProcessLookupError:
            pass


async def producer(stop, args, pool):
    """Bursty producer: publish a spike of tasks, then pause, repeat. Note: real
    Celery .delay() is the speed ceiling (~1-2k/s single-thread), so a large spike
    is smeared over a few seconds rather than instantaneous - use sawtooth.py for
    sharp spikes."""
    loop = asyncio.get_running_loop()
    i = 0
    while not stop.is_set():
        produced = 0
        while produced < args.produce_burst and not stop.is_set():
            chunk = min(500, args.produce_burst - produced)
            i = await loop.run_in_executor(pool, publish_batch, i, chunk, args.payload_items)
            produced += chunk
        print(f"[produce] spike of {produced:,}")
        if await _sleep_or_stop(stop, args.produce_pause):  # pause between spikes
            break


async def start_worker(args):
    logf = open(WORKER_LOG, "w")
    return await asyncio.create_subprocess_exec(
        sys.executable,
        "-m",
        "celery",
        "-A",
        "celery_app",
        "worker",
        "--pool=solo",
        "-Q",
        args.queue,
        "--loglevel=INFO",
        "--without-gossip",
        "--without-mingle",
        "--without-heartbeat",
        "--prefetch-multiplier=1",  # keep the backlog in Dragonfly, not the worker
        cwd=HERE,
        start_new_session=True,  # shield it from terminal Ctrl-C; we drive it
        stdout=logf,
        stderr=logf,
    )


async def consumer(stop, args, worker):
    await asyncio.sleep(args.worker_boot)  # let the worker connect first
    if worker.returncode is not None:
        print(f"[worker] exited early (rc={worker.returncode}); see {WORKER_LOG}")
        return
    _signal_worker(worker, signal.SIGSTOP)
    print("[worker] paused; queue will build")
    while not stop.is_set():
        if await _sleep_or_stop(stop, args.drain_every):  # build-up phase
            break
        _signal_worker(worker, signal.SIGCONT)
        print("[worker] resumed - draining")
        if await _sleep_or_stop(stop, args.drain_window):
            break
        _signal_worker(worker, signal.SIGSTOP)
        print("[worker] paused")


async def monitor(stop, r, args):
    while not stop.is_set():
        llen = await r.llen(args.queue)
        used = (await r.info("memory")).get("used_memory", 0)
        print(f"[mon]  LLEN({args.queue})={llen:>9,}  used_memory={used / 1024 / 1024:6.1f} MiB")
        if await _sleep_or_stop(stop, args.monitor_every):
            break


async def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--host", default="localhost")
    ap.add_argument("--port", type=int, default=6379)
    ap.add_argument("--queue", default="celery")
    ap.add_argument(
        "--produce-burst",
        type=int,
        default=20000,
        help="tasks per producer spike (real .delay() is the speed ceiling)",
    )
    ap.add_argument(
        "--produce-pause",
        type=float,
        default=5.0,
        help="seconds the producer pauses between spikes (0 = continuous)",
    )
    ap.add_argument(
        "--payload-items",
        type=int,
        default=0,
        help="structured records per task: 0=~1KB, 50=~12KB, 100=~22KB each",
    )
    ap.add_argument(
        "--drain-every", type=float, default=20.0, help="build-up seconds between worker wake-ups"
    )
    ap.add_argument(
        "--drain-window", type=float, default=8.0, help="seconds the worker drains on each wake-up"
    )
    ap.add_argument(
        "--worker-boot",
        type=float,
        default=5.0,
        help="seconds to let the worker boot before first pause",
    )
    ap.add_argument("--monitor-every", type=float, default=2.0)
    args = ap.parse_args()

    r = aioredis.Redis(host=args.host, port=args.port, db=0)
    pool = ThreadPoolExecutor(max_workers=1)  # serialize real .delay() calls
    stop = asyncio.Event()
    loop = asyncio.get_running_loop()
    for sig in (signal.SIGINT, signal.SIGTERM):
        loop.add_signal_handler(sig, stop.set)

    print(
        f"real Celery: producer spikes of {args.produce_burst:,} (pause {args.produce_pause}s); "
        f"worker drains {args.drain_window}s every {args.drain_every}s. "
        f"Worker log: {WORKER_LOG}. Ctrl-C to stop."
    )
    worker = await start_worker(args)
    try:
        await asyncio.gather(
            producer(stop, args, pool), consumer(stop, args, worker), monitor(stop, r, args)
        )
    finally:
        # Resume (so it can act on terminate), then shut the worker down cleanly.
        _signal_worker(worker, signal.SIGCONT)
        if worker.returncode is None:
            worker.terminate()
            try:
                await asyncio.wait_for(worker.wait(), timeout=10)
            except asyncio.TimeoutError:
                worker.kill()
        pool.shutdown(wait=False)
        try:
            await r.aclose()
        except AttributeError:
            await r.close()
    print("stopped.")


if __name__ == "__main__":
    asyncio.run(main())

File: tools/sidekiq_zstd_experiment/churn.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Fully-real continuous spike-and-drain load for the ZSTD-dict peak-memory
# experiment, driven from one Ruby process:
#
#   * producer (thread): bursts of real HardWorker.perform_bulk -> LPUSH
#     queue:default, then pauses, repeat.
#   * consumer (thread): drives a REAL `sidekiq` worker subprocess. The worker is
#     paused with SIGSTOP so the queue builds, then every --drain-every seconds it
#     is resumed (SIGCONT) for a --drain-window burst during which it actually
#     BRPOPs + executes jobs (decompressing nodes on read), then paused again.
#   * monitor (thread): prints LLEN(queue:default) + used_memory.
#
# Runs until Ctrl-C. The thread-local ZSTD dictionary stays trained across drains,
# so each rebuild compresses immediately - realistic steady-state behavior.
#
# Usage:
#   bundle exec ruby churn.rb --produce-burst 20000 --produce-pause 5 \
#                             --payload-items 50 --drain-every 20 --drain-window 8

require_relative "worker"
require "optparse"
require "timeout"

$stdout.sync = true

WORKER_LOG = "/tmp/sidekiq_zstd_worker.log"

opts = {
  queue: "default",
  produce_burst: 20_000,
  produce_pause: 5.0,
  payload_items: 0,
  drain_every: 20.0,
  drain_window: 8.0,
  worker_boot: 5.0,
  concurrency: 10,
  monitor_every: 2.0,
}

OptionParser.new do |o|
  o.banner = "Usage: churn.rb [options]"
  o.on("--queue NAME", String, "Sidekiq queue name (list key is queue:NAME)") { |v| opts[:queue] = v }
  o.on("--produce-burst N", Integer, "jobs per producer spike") { |v| opts[:produce_burst] = v }
  o.on("--produce-pause SECS", Float, "pause between spikes (0 = continuous)") { |v| opts[:produce_pause] = v }
  o.on("--payload-items N", Integer,
       "structured records per job: 0=~1KB, 50=~12KB, 100=~22KB each") { |v| opts[:payload_items] = v }
  o.on("--drain-every SECS", Float, "build-up seconds between worker wake-ups") { |v| opts[:drain_every] = v }
  o.on("--drain-window SECS", Float, "seconds the worker drains on each wake-up") { |v| opts[:drain_window] = v }
  o.on("--worker-boot SECS", Float, "seconds to let the worker boot before first pause") { |v| opts[:worker_boot] = v }
  o.on("--concurrency N", Integer, "sidekiq worker threads (drain speed)") { |v| opts[:concurrency] = v }
  o.on("--monitor-every SECS", Float, "seconds between monitor prints") { |v| opts[:monitor_every] = v }
end.parse!

QUEUE_KEY = "queue:#{opts[:queue]}"

# --- cooperative stop, with snappy Ctrl-C via a self-pipe -------------------
@mutex = Mutex.new
@cond = ConditionVariable.new
@stop = false

def stop?
  @mutex.synchronize { @stop }
end

def signal_stop
  @mutex.synchronize do
    @stop = true
    @cond.broadcast
  end
end

# Sleep up to `secs`, but wake immediately once stop is set. Returns true if stop
# was signalled (so callers can `break if sleep_or_stop(...)`).
def sleep_or_stop(secs)
  deadline = Time.now + secs
  @mutex.synchronize do
    until @stop
      remaining = deadline - Time.now
      break if remaining <= 0

      @cond.wait(@mutex, remaining)
    end
    @stop
  end
end

def signal_worker(pid, sig)
  Process.kill(sig, pid)
rescue Errno::ESRCH
  # worker already gone
end

def spawn_worker(opts)
  log = File.open(WORKER_LOG, "w")
  cmd = ["bundle", "exec", "sidekiq",
         "-r", File.join(__dir__, "worker.rb"),
         "-q", opts[:queue],
         "-c", opts[:concurrency].to_s]
  launch = lambda do
    Process.spawn(
      { "REDIS_URL" => REDIS_URL }, *cmd,
      pgroup: true,         # own process group: shield from terminal Ctrl-C; we drive it
      out: log, err: log, chdir: __dir__
    )
  end
  # If we were started via `bundle exec`, spawn the child with a clean bundler env
  # so it resolves this Gemfile rather than inheriting a nested context.
  defined?(Bundler) ? Bundler.with_unbundled_env(&launch) : launch.call
end

def producer_loop(opts)
  i = 0
  until stop?
    produced = 0
    while produced < opts[:produce_burst] && !stop?
      chunk = [1_000, opts[:produce_burst] - produced].min
      batch = Array.new(chunk) do |j|
        idx = i + j
        ["b3e4b923-8a77-4053-aff0-#{100_000 + idx}", make_payload(idx, opts[:payload_items])]
      end
      HardWorker.perform_bulk(batch)
      i += chunk
      produced += chunk
    end
    puts format("[produce] spike of %d", produced)
    break if sleep_or_stop(opts[:produce_pause]) # pause between spikes
  end
end

def consumer_loop(worker_pid, opts)
  sleep_or_stop(opts[:worker_boot]) # let the worker connect first
  return if stop?

  signal_worker(worker_pid, "STOP")
  puts "[worker] paused; queue will build"
  until stop?
    break if sleep_or_stop(opts[:drain_every]) # build-up phase

    signal_worker(worker_pid, "CONT")
    puts "[worker] resumed - draining"
    break if sleep_or_stop(opts[:drain_window])

    signal_worker(worker_pid, "STOP")
    puts "[worker] paused"
  end
end

def monitor_loop(opts, queue_key)
  until stop?
    llen = Sidekiq.redis { |c| c.call("LLEN", queue_key) }
    info = Sidekiq.redis { |c| c.call("INFO", "memory") }
    used = (info[/used_memory:(\d+)/, 1] || 0).to_i
    puts format("[mon]  LLEN(%s)=%9d  used_memory=%6.1f MiB", queue_key, llen, used / 1024.0 / 1024.0)
    break if sleep_or_stop(opts[:monitor_every])
  end
end

# --- main -------------------------------------------------------------------
self_read, self_write = IO.pipe
%w[INT TERM].each { |sig| trap(sig) { self_write.puts(sig) } }

puts format(
  "real Sidekiq: producer spikes of %d (pause %ss); worker drains %ss every %ss. " \
  "Worker log: %s. Ctrl-C to stop.",
  opts[:produce_burst], opts[:produce_pause], opts[:drain_window], opts[:drain_every], WORKER_LOG
)

worker_pid = spawn_worker(opts)
threads = [
  Thread.new { producer_loop(opts) },
  Thread.new { consumer_loop(worker_pid, opts) },
  Thread.new { monitor_loop(opts, QUEUE_KEY) },
]

# Block until a signal arrives (or all workers finish on their own).
watcher = Thread.new do
  self_read.gets
  signal_stop
end
threads.each(&:join)
signal_stop
watcher.kill

# Resume (so it can act on TERM), then shut the worker down cleanly.
signal_worker(worker_pid, "CONT")
signal_worker(worker_pid, "TERM")
begin
  Timeout.timeout(15) { Process.wait(worker_pid) }
rescue Timeout::Error
  signal_worker(worker_pid, "KILL")
  Process.wait(worker_pid)
rescue Errno::ECHILD
  # already reaped
end
puts "stopped."
